src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep.java

```diff
@@ -0,0 +1,217 @@
+        this.testResults = testResults;
+    }
+    
+    public String getTestResults() {
+        return testResults;
+    }
+
+    public boolean isKeepLongStdio() {
+        return keepLongStdio;
+    }
+
+    @DataBoundSetter
+    public void setKeepLongStdio(boolean keepLongStdio) {
+        this.keepLongStdio = keepLongStdio;
+    }
+
+    @Nonnull
+    public List<TestDataPublisher> getTestDataPublishers() {
+        return testDataPublishers == null ? Collections.<TestDataPublisher>emptyList() : testDataPublishers;
+    }
+
+    @DataBoundSetter
+    public void setTestDataPublishers(@Nonnull List<TestDataPublisher> testDataPublishers) {
+        this.testDataPublishers = testDataPublishers;
+    }
+
+    public double getHealthScaleFactor() {
+        return healthScaleFactor == null ? 1.0 : healthScaleFactor;
+    }
+
+    @DataBoundSetter
+    public void setHealthScaleFactor(double healthScaleFactor) {
+        this.healthScaleFactor = Math.max(0.0, healthScaleFactor);
+    }
+
+    public boolean isAllowEmptyResults() {
+        return allowEmptyResults;
+    }
+
+    @DataBoundSetter
+    public void setAllowEmptyResults(boolean allowEmptyResults) {
+        this.allowEmptyResults = allowEmptyResults;
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        JUnitResultArchiver delegate = new JUnitResultArchiver(testResults);
+        delegate.setAllowEmptyResults(allowEmptyResults);
+        delegate.setHealthScaleFactor(getHealthScaleFactor());
+        delegate.setKeepLongStdio(keepLongStdio);
+        delegate.setTestDataPublishers(getTestDataPublishers());
+        return new Execution(context, delegate);
+    }
+
+    static class Execution extends StepExecution {
+
+        private final JUnitResultArchiver archiver;
+
+        Execution(StepContext context, JUnitResultArchiver archiver) {
+            super(context);
+            this.archiver = archiver;
+        }
+
+        @Override
+        public boolean start() throws Exception {
+            Run<?, ?> r = getContext().get(Run.class);
+            String id = getContext().get(FlowNode.class).getId();
+            r.addAction(new PipelineRealtimeTestResultAction(id, getContext().get(FilePath.class), archiver.isKeepLongStdio(), archiver.getTestResults()));
+            AbstractRealtimeTestResultAction.saveBuild(r);
+            getContext().newBodyInvoker().withCallback(new Callback(id, archiver)).start();
+            return false;
+        }
+
+        private static final long serialVersionUID = 1;
+
+    }
+
+    static class Callback extends BodyExecutionCallback.TailCall {
+
+        private final String id;
+        private final JUnitResultArchiver archiver;
+
+        Callback(String id, JUnitResultArchiver archiver) {
+            this.id = id;
+            this.archiver = archiver;
+        }
+
+        @Override
+        protected void finished(StepContext context) throws Exception {
```

Comment:
Note that `TailCall` does not indicate whether the body succeeded or failed. If the latter, it is possible that there will not have been any test results produced so far; if `allowEmptyResults` has not been enabled, and the build’s `result` is still null (typical for Pipeline), `archiver.perform` could throw `AbortException`. I doubt this is really a problem because that error will merely be `addSuppressed` to the main build failure: at most it will be displayed at the end of a stack trace, but it cannot on its own break the build.

If desired, would be fixable by extending `BodyExecutionCallback` directly and in the case of `onFailure`, calling `archiver.setAllowEmptyResults(true)`. That will cause the archiver to always print a message to the build log, but not add a suppressed exception.